Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+    start_at = 1          # 1 in the paper
+    number_training_steps = 10**8 # It should start doing well after 1e6??
+    Print_progress_after = 10 ** 2
+    Copy_model_after = 10 ** 4  # 1e4 in the blog?
+
+
+    number_random_actions = args.num_rand_acts # Should be at least 36. 5e4 in the paper?
+    save_model_after_steps = args.save_after # Try 1e5?
+    mem_size = args.mem_size # 1e6 in the paper?
+
+    print('num_rand_acts = ', number_random_actions, ', save_after = ', save_model_after_steps, ', mem_size = ', mem_size)
+
+    # Make the model
+    model = atari_model_mask()
+    model.summary()
+
+    # Make the memories
+    mem_frames = RingBuf_simple(mem_size)
+    mem_actions = RingBuf_simple(mem_size)
+    mem_rewards = RingBuf_simple(mem_size)
+    mem_is_terminal = RingBuf_simple(mem_size)
+
+    print('Setting up Breakout and pre-filling memory with random actions...')
+
+    # Create and reset the Atari env, and process the initial screen:
+    env = gym.make('BreakoutDeterministic-v4')
+    # Probably don't need these 3 lines:
+    env.reset()
+    prev_screen = env.render(mode='rgb_array')
+    preproc_screen = preprocess(prev_screen)
+
+    ## First make some random actions, and initially fill the memories with these
+    for i in range(number_random_actions+1):
+
+        iteration = i
+
+        # Random action
+        action = env.action_space.sample()
+        new_frame, reward, is_terminal, _ = env.step(action)
+
+        reward = transform_reward(reward)  # Make reward just +1 or -1
+        new_frame = preprocess(new_frame)  # Preprocess frame before saving it
+
+        # Add to memory
+        mem_frames.append(new_frame, iteration)  # The frame we ended up in after the action
+        mem_actions.append(action, iteration)  # The action we did
+        mem_rewards.append(reward, iteration)  # The reward we received after doing the action
+        mem_is_terminal.append(is_terminal, iteration)  # Whether or not the new frame is terminal (assuming is_terminal is the same as is_terminal)
+
+
+    # Now do actions using the DNN, and train as we go...
+    print('Finished the', number_random_actions, 'random actions...')
+    tic=0
+
+    for i in range(number_training_steps):
+
+        iteration = number_random_actions+i
+
+        # Copy model every now and then and fit to this: makes it more stable
+        if np.mod(i, Copy_model_after) == 0:
+            target_model = copy_model(model)
+
+        action, reward, is_terminal, epsilon = q_iteration(env, model, target_model, iteration, mem_frames, mem_actions, mem_rewards, mem_is_terminal, mem_size)
+
+        # Print progress, time, and SAVE the model
+        if np.mod(i+1, Print_progress_after) == 0:
+            #print('Training steps done: ', i + 1, ', Action = ', action, ', Reward = ', reward, ', Terminal = ',
+            # is_terminal, ', Epsilon', epsilon)
+            print('Training steps done: ', i+1,', Epsilon',epsilon)
```

Comment:
This is fine but you might find Python's string format useful: https://pyformat.info/ and https://docs.python.org/3/library/string.html#string-formatting